File: 6. Spiral Traversal.py
"""Topic: Spiral Traversal"""
def spiral(A):
    top,left=0,0
    bottom=len(A)-1; right=len(A[0])-1
    out=[]
    while top<=bottom and left<=right:
        for j in range(left,right+1): out.append(A[top][j])
        top+=1
        for i in range(top,bottom+1): out.append(A[i][right])
        right-=1
        if top<=bottom:
            for j in range(right,left-1,-1): out.append(A[bottom][j])
            bottom-=1
        if left<=right:
            for i in range(bottom,top-1,-1): out.append(A[i][left])
            left+=1
    return out

if __name__=='__main__':
    A=[[1,2,3],[4,5,6],[7,8,9]]
    print(*spiral(A))

# TC: O(n*m)
# SC: O(n*m) output

File: 6. Spiral Traversal.cpp
// Topic: Spiral Traversal
#include <bits/stdc++.h>
using namespace std;
int main()
{
    vector<vector<int>> A = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
    int top = 0, left = 0, bottom = A.size() - 1, right = A[0].size() - 1;
    vector<int> out;
    while (top <= bottom && left <= right)
    {
        for (int j = left; j <= right; ++j) out.push_back(A[top][j]);
        ++top;
        for (int i = top; i <= bottom; ++i) out.push_back(A[i][right]);
        --right;
        if (top <= bottom)
        {
            for (int j = right; j >= left; --j) out.push_back(A[bottom][j]);
            --bottom;
        }
        if (left <= right)
        {
            for (int i = bottom; i >= top; --i) out.push_back(A[i][left]);
            ++left;
        }
    }
    for (auto v : out) cout << v << " ";
    cout << "\n";
}

// TC: O(n*m)
// SC: O(1) extra (output excluded)
